### PR TITLE
Add kickstart %pre-install section support

### PR DESCRIPTION
**Pull request by bcl**
The %pre-install section will run scripts after all the filesystems have
been setup and mounted on /mnt/sysimage, but before any packages have
been installed.

### CODE REVIEW

**Comment by cgwalters** [comment]
What's the use case?

File: pyanaconda/kickstart.py [hunk 1]
```diff
@@ -2121,6 +2123,16 @@ def runPreScripts(scripts):
 
     log.info("All kickstart %%pre script(s) have been run")
 
+def runPreInstallScripts(scripts):
+    preInstallScripts = [s for s in scripts if s.type == KS_SCRIPT_PREINSTALL]
+
+    if len(preInstallScripts) == 0:
```
Comment (vpodzime) [inline pyanaconda/kickstart.py]: ``if not any(preInstallScripts):`` would do the same, so in case you (too) like it better, you can use it here.

**Comment by vpodzime** [comment]
Same comment/concern/question here as in the [related pykickstart PR] (https://github.com/rhinstaller/pykickstart/pull/10)

**Comment by clumens** [comment]
Colin - Check out https://fedorahosted.org/fpc/ticket/508.  Basically there's no way for people to muck around with the target system until after packages have been installed, because we don't expose anything between filesystems being made and packages being installed.  We went around and around in irc one day about how to best provide a way for them to do the UID things they need, and determined anaconda had no way to help them yet.  This would take care of it.

File: pyanaconda/kickstart.py (hunk 1 above)
Comment (bcl) [inline pyanaconda/kickstart.py]: I like len() much better, thanks :)

**Comment by cgwalters** [comment]
@clumens Thanks!  That's exactly the info I was looking for (and I think should be in the commit message, so if github goes down someone can find it later).

On the rpm-ostree side we support doing something very similar to this now, see https://github.com/projectatomic/rpm-ostree/pull/56  (TL;DR, we support injecting `passwd` and `group` files before the RPMs get installed).

**Comment by cgwalters** [comment]
One issue that has always existed with `%post --nochroot` and will also be present for this new `%pre-install` is the question of "what commandline tools are available" - in practice, I think script writers will simply test using whatever version of Anaconda (i.e. lorax templates) and use what's available.  Probably most of them are just executing things from `coreutils` and the like.

But it does mean that over time we'll have to consider external scripts when changing the lorax templates and likely be more conservative about removals.

(Random alternatives would be being able to pull down a Docker container that contains the user script, or having the user script come as a package, and doing a second install root...but probably the sanest thing is to just document "Try to rely only on core packages such as `coreutils`?)

**Comment by jasontibbitts** [comment]
Just came across this; I was agitating for something like this on IRC a while back and I'm glad this popped up as I'm sure my incomplete implementation would have been horrible anyway.

Regarding cgwalters's comment above, how is this any different than scripts in %pre?  It should be running with exactly the same set of available tools, just slightly later in the install process.  How is compatibility maintained with %pre scripts across anaconda releases?  I haven't been able to find any guarantee about what is available there.

I think if it were simply stated somewhere that a fixed set of utilities will be guaranteed to be available and other things may come and go, that should be good enough.  People doing things at this level should be able to deal with that kind of thing, and if someone has a use case for a particular utility being available, they can always make a case for moving it into the guaranteed set.

**Comment by opoplawski** [comment]
My two use cases:
- cobbler copies the %pre output logs onto the target system after initialized
- need to modify /etc/login.defs, /etc/rpm/macros on target before install begins

**Comment by bcl** [comment]
> Regarding cgwalters's comment above, how is this any different than scripts in %pre? It should be running with exactly the same set of available tools, just slightly later in the install process. How is compatibility maintained with %pre scripts across anaconda releases? I haven't been able to find any guarantee about what is available there.

There's really no guarantee, but we do try to take into consideration kickstart users :) We also include things that are useful during rescue which will have some overlap. If there is something you need, feel free to file a request against lorax.

This has been pushed and will be in Fedora 23.

**Pull request closed without merging**